File: 1962. Remove Stones to Minimize the Total/Solution.java
class Solution {
    public int minStoneSum(int[] piles, int k) {
        int res = 0;
        int[] counts = new int[10001];
        for (int p : piles) counts[p]++;
        for (int i = 10000; i > 1 && k > 0; i--) {
            if (counts[i] == 0) continue;
            counts[floor(i)] += Math.min(k, counts[i]);
            if (k >= counts[i]) {
                k -= counts[i];
                counts[i] = 0;
            } else {
                counts[i] -= k;
                k = 0;
            }
        }
        for (int i = 0; i < 10001; i++) {
            if (counts[i] == 0) continue;
            res += i * counts[i];
        }
        return res;
    }
    public int floor(int x) {
        return (int) ((x + 1) / 2);
    }
}

File: 1962. Remove Stones to Minimize the Total/Solution.py
class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        counts = [0 for i in range(10001)]
        for p in piles:
            counts[p] += 1
        for i in range(10000, 1, -1):
            if counts[i] == 0: continue
            counts[self.floor(i)] += min(k, counts[i])
            if k >= counts[i]:
                k -= counts[i]
                counts[i] = 0
            else:
                counts[i] -= k
                k = 0
            if k == 0: break
        return sum([i * counts[i] for i in range(10001)])

    def floor(self, x):
        return (x + 1) // 2
